src/attr/_make.py

```diff
@@ -1864,7 +1865,46 @@ def _add_eq(cls, attrs=None):
     return cls
 
 
-_already_repring = threading.local()
+# Thread-local global to track attrs instances which are already being repr'd.
+# This is needed because there is no other (thread-safe) way to pass info
+# about the instances that are already being repr'd through the call stack
+# in order to ensure we don't perform infinite recursion.
+#
+# For instance, if an instance contains a dict which contains that instance,
+# we need to know that we're already repr'ing the outside instance from within
+# the dict's repr() call.
+_repr_context = threading.local()
+
+
+@contextlib.contextmanager
+def _is_already_repring(inst):
+    """
+    Checks whether the given instance of this class is already being repr'd.
+    If not, sets up the thread-local context to record that we are repr'ing
+    this instance.
+
+    Yields ``True`` if we're already repr'ing this class;
+    ``False`` if we are not already repr'ing it.
+    """
+    try:
+        working_set = _repr_context.working_set
+    except AttributeError:
+        # Since 'self' remains on the stack (i.e.: strongly referenced)
```

Comment:
Do you mean `inst`?

Reply:
After much consideration, I have decided that I do.